### PR TITLE
Quest to determine type of recycling amenity

### LINKED ISSUE #222: amenity=recycling without recycling_type tag
`amenity=recycling` can be a centre or only a container. Without it `recycling_type=*` the poi is rather useless.

- `recycling_type=centre`: a recycling centre.
- `recycling_type=container`: a recycling container.
- `recycling_type=container` and `location=underground` for an underfloor-container

### PR DESCRIPTION
**Pull request by Teester**
This pull request fixes #222.

It searches for amenity=recycling without recycling_type and asks:
"What type of recycling amenity is this?"

Available answers:
- "An above ground recycling container" - tags recycling_type=container
- "An underground recycling container"- tags recycling_type=container & location=underground
- "A recycling centre" - tags recycling_type=centre
- "Can't say" - standard options

Image:
screenshot_20170703-112552

Possible Issues:
- Is text only suitable?
- 3 possible answers just about triggers scrolling in the answer page for me using a Moto G.  Is this acceptable.
- Is th ordering of the answers suitable?

### CODE REVIEW

**Comment by westnordost** [comment]
Hmm, do _you_ like it? To me, it sounds a bit wordy. I.e. what is an "recycling amenity"? Perhaps like this (should fit into one line):
```
"What recycling place is this?"
[icon]         [icon]            [icon]
container     container     recycling center
            (underground)
```

**Comment by westnordost** [comment]
I could paint icons for these if that makes you go with pictures :-)

**Comment by Teester** [comment]
Perhaps 'facility' rather than 'amenity' would work better.

I've tested out using images and text as suggested. rather than text alone.  It definitely looks better, but space is at a premium using 3 buttons in a row and I've had to mess with the layout weights to prevent text from wrapping where it shouldn't.  The example uses ic_bicycle.xml as a placeholder icon. 
I would appreciate if you could create suitable icons.

screenshot_20170704-011700

**Comment by westnordost** [comment]
Right, facility is better.

I am not sure regarding the layout, I think it should not be done like this. It is important that the UI behaves consistently throughout the app.  I'd say the expectation of the user is that only the **OK** and possible other buttons in the bottom bar will result in the quest being solved immediately.
From the sport quest, paved quest, roof shape quest etc. they are used that they first select something, then click okay. I'm not sure if making the buttons not borderless will make it better.

Does anything speak against having the same UI as i.e. for roof shapes? I think it shouldn't even be necessary to use a RecyclerView with the ImageSelectAdapter.

Anyway, I started making icons but in the end thought that actual photos would be nicer. What do you think? (this is for xxhdpi):
recycling_type_container
recycling_type_centre
recycling_type_container_underground

**Comment by Teester** [comment]
A couple of options:
recycling

The one on the right is my favourite.  

The images don't look as good in the left one and regarding the middle one, I don't think that "Select one option:" is necessary as the dialog will only let you select one option anyway.

Regarding the images, I'm not sure the underground container one is obvious, but I haven't seen any better images anywhere and I think that the text overlaid on the image makes things clear.  In any case, location=underground and recycling_type=container is rare in OSM.

There are currently conflicts in the branch, but I don't plan to update it until we have consensus on how the quest should look.

**Comment by westnordost** [comment]
If you are using the pre-made ImageSelect-thingy that has the "Select one" built in, I'd say simply use that. It's good to have consistent UI. Otherwise, the right one is also fine.

**Comment by westnordost** [comment]
This is basically ready to merge. However, I want to refactor something before because the ImageSelectQuestAnswerFragment is not flexible enough but currently have no time for that so I have to do this later.

As far as you are concerned, the quest implementation is done.

**Comment by westnordost** [comment]
Good?

**Comment by Teester** [comment]
Looks good to me.  Everything still works correctly.  This will make ImageListQuestAnswerFragment much more adaptable for future quests.

The tests in AddRecyclingTypeTest were failing due to adding strings rather than arraylist values to AddRecyclingTypeForm.OSM_VALUES.  I've fixed that in the most recent commit.

**Comment by westnordost** [comment]
Today it came to me that perhaps the UI flow of quest should be improved because I found out that `amenity=recycling` nodes have a strong tendency to be containers and ways to be recycling centres.

In numbers, it looks like roughly over 95-99% of `amenity=recycling` **ways** are already tagged as recycling centres.
For **nodes** on the other hand, it is something like 50% which are tagged as containers (and the rest not).

So, the quest misses the point a bit, as it presents an option (recycling centre) that a user will almost never choose (and then, most likely, almost always on an area and not on a node).

Based on this data, perhaps it would be better to ask the user:
* _Is this recycling container one of these fancy underground containers?_ for any recycling node, assuming that it will be a container and adding an option in "other answers" that says "it is actually a recycling center"
* And then perhaps ask the user whether a recycling way is a recycling center. But there are so few elements worldwide that this is not worth it, IMO

**Comment by HolgerJeromin** [comment]
Very good idea. 
Do we have the size of the area to ask the container question for very small things, too?
Perhaps this is not needed...

**Comment by goldfndr** [comment]
I was checking tags used by StreetComplete (after #1267 was merged). Shouldn't this quest use location=surface instead of location=overground? (Or perhaps at least distinguish between "above ground" and "surface"?)

**Comment by westnordost** [comment]
Overground is used much more often with the[ `location` key](https://taginfo.openstreetmap.org/keys/location#overview).

**Comment by goldfndr** [comment]
A testament to StreetComplete's popularity.

* Check the location=overground tag at http://taghistory.raifer.tech/? with only nodes (not ways+relations) and you'll notice the spike starts in late August (corresponding to [v1.2's release](https://github.com/westnordost/StreetComplete/releases/tag/v1.2)).
* The location=overground tag is used for nearly 34000 ways, of which 31000 are man_made=pipeline. It's used for 9300 nodes, of which slightly over 8400 are amenity=recycling and recycling=container.

**Comment by matkoniecz** [comment]
Note that `location=surface` is even more StreetComplete-inspired tagging. I invented and documented it to allow implementing #1267 - see https://github.com/westnordost/StreetComplete/issues/907#issuecomment-368252490

**Comment by westnordost** [comment]
Hmm. What now?

Did you also clarify/change the definition of 'overground' when you added 'surface'?

Am 30. November 2018 14:05:37 MEZ schrieb Mateusz Konieczny <notifications@github.com>:
>Note that `location=surface` is even more StreetComplete-inspired
>tagging. I invented and documented it to allow implementing #1267 - see
>https://github.com/westnordost/StreetComplete/issues/907#issuecomment-368252490
>

**Comment by matkoniecz** [comment]
@westnordost From history of the page I see that my sole edit was to add 

> *{{Tag|location||surface}} Feature is on surface of the ground

https://wiki.openstreetmap.org/w/index.php?title=Key:location&diff=1574512&oldid=1542678

At that time `location=overground` was described as 

> location=overground Used where a feature such as a pipeline or cable is over ground, located close to the surface, not allowing a person or vehicle to pass underneath.

so I have seen no need to change it
